File: kernel/probes/tracepoints.c
// SPDX-License-Identifier: GPL-2.0-only
// Copyright (C) 2026 rezky_nightky

//! Tracepoints probe.
//!
//! Discovered facts:
//!   tracepoints.count       — number of registered tracepoints
//!   tracepoints.subsystems  — comma-separated subsystem names
//!                             (best-effort static list)

#include <linux/kernel.h>
#include <linux/string.h>
#include <linux/tracepoint.h>
#include "zenvecha.h"

static char count_buf[16];
static char subsystems_buf[128] = "sched,block,net,irq,syscalls,fs,mm";

static const struct capability_descriptor descriptors[] = {
	{ .key = "tracepoints.count",      .value = count_buf      },
	{ .key = "tracepoints.subsystems", .value = subsystems_buf },
};

static u64 tracepoint_count;

static void count_tracepoint(struct tracepoint *tp, void *priv)
{
	(void)tp;
	(void)priv;
	tracepoint_count++;
}

const struct capability_descriptor *tracepoints_probe_discover(void)
{
	tracepoint_count = 0;

#ifdef CONFIG_TRACEPOINTS
	for_each_kernel_tracepoint(count_tracepoint, NULL);
#endif

	snprintf(count_buf, sizeof(count_buf), "%llu", tracepoint_count);

	return descriptors;
}

size_t tracepoints_probe_count(void)
{
	return ARRAY_SIZE(descriptors);
}

File: src/core/caps/kernel_cap/tracepoints.rs
// Copyright (C) 2026 rezky_nightky
// SPDX-License-Identifier: GPL-3.0-only

//! Tracepoint capability probes.
//!
//! Uses: kernel_text()

use crate::core::capability::Capability;
use crate::core::caps::kernel_cap::kernel_text;
use crate::core::evidence::Evidence;

/// Total tracepoint count available in the kernel.
pub struct KernelTracepointCount;

impl Capability for KernelTracepointCount {
    fn id(&self) -> &'static str {
        "kernel.tracepoints.count"
    }
    fn label(&self) -> &'static str {
        "Tracepoint Count (module)"
    }
    fn probe(&self) -> Evidence {
        // /proc/zenvecha/tracepoints/count → e.g. "1427"
        kernel_text(self.id(), "tracepoints.count")
    }
}

/// Tracepoint subsystems available (e.g. "sched,block,net,irq").
pub struct KernelTracepointSubsystems;

impl Capability for KernelTracepointSubsystems {
    fn id(&self) -> &'static str {
        "kernel.tracepoints.subsystems"
    }
    fn label(&self) -> &'static str {
        "Tracepoint Subsystems (module)"
    }
    fn probe(&self) -> Evidence {
        // /proc/zenvecha/tracepoints/subsystems → "sched,block,net,..."
        kernel_text(self.id(), "tracepoints.subsystems")
    }
}
